Forwarded: https://github.com/BuddiesOfBudgie/budgie-desktop/pull/70
Author: Evan Maddock <maddock.evan@vivaldi.net>
Last-Update: 2022-03-15
Description: [PATCH] Improve	racyness when a	notification is	closed and
 another one shown in	quick succession

Signed-off-by: Evan Maddock <maddock.evan@vivaldi.net>
---
 src/daemon/notifications/dbus.vala  | 47 +++++++++++++----------------
 src/daemon/notifications/popup.vala |  3 ++
 2 files changed, 24 insertions(+), 26 deletions(-)

diff --git a/src/daemon/notifications/dbus.vala b/src/daemon/notifications/dbus.vala
index 984ff759..867815b8 100644
--- a/src/daemon/notifications/dbus.vala
+++ b/src/daemon/notifications/dbus.vala
@@ -291,8 +291,7 @@
 			Gdk.Rectangle rect = mon.get_geometry();
 
 			/* Set the x, y position of the notification */
-			int x;
-			int y;
+			int x = 0, y = 0;
 			calculate_position(popup, rect, out x, out y);
 			popup.move(x, y);
 		}
@@ -304,27 +303,29 @@
 		private void calculate_position(Popup window, Gdk.Rectangle rect, out int x, out int y) {
 			var pos = (NotificationPosition) this.panel_settings.get_enum("notification-position");
 			var latest = this.popups.get(this.latest_popup_id);
+			bool latest_exists = false;
+			int existing_height = 0, existing_x = 0, existing_y = 0;
+
+			if (latest != null && !latest.destroying) {
+				latest_exists = true;
+				existing_height = latest.get_child().get_allocated_height();
+				latest.get_position(out existing_x, out existing_y);
+			}
 
 			switch (pos) {
 				case NotificationPosition.TOP_LEFT:
-					if (latest != null) { // If a notification is already being displayed
-						int nx;
-						int ny;
-						latest.get_position(out nx, out ny);
-						x = nx;
-						y = ny + latest.get_child().get_allocated_height() + BUFFER_ZONE;
+					if (latest_exists) { // If a notification is already being displayed
+						x = existing_x;
+						y = existing_y + existing_height + BUFFER_ZONE;
 					} else { // This is the first nofication on the screen
 						x = rect.x + BUFFER_ZONE;
 						y = rect.y + INITIAL_BUFFER_ZONE;
 					}
 					break;
 				case NotificationPosition.BOTTOM_LEFT:
-					if (latest != null) { // If a notification is already being displayed
-						int nx;
-						int ny;
-						latest.get_position(out nx, out ny);
-						x = nx;
-						y = ny - latest.get_child().get_allocated_height() - BUFFER_ZONE;
+					if (latest_exists) { // If a notification is already being displayed
+						x = existing_x;
+						y = existing_y - existing_height - BUFFER_ZONE;
 					} else { // This is the first nofication on the screen
 						x = rect.x + BUFFER_ZONE;
 
@@ -334,12 +335,9 @@
 					}
 					break;
 				case NotificationPosition.BOTTOM_RIGHT:
-					if (latest != null) { // If a notification is already being displayed
-						int nx;
-						int ny;
-						latest.get_position(out nx, out ny);
-						x = nx;
-						y = ny - latest.get_child().get_allocated_height() - BUFFER_ZONE;
+					if (latest_exists) { // If a notification is already being displayed
+						x = existing_x;
+						y = existing_y - existing_height - BUFFER_ZONE;
 					} else { // This is the first nofication on the screen
 						x = (rect.x + rect.width) - NOTIFICATION_WIDTH;
 						x -= BUFFER_ZONE; // Don't touch edge of the screen
@@ -351,12 +349,9 @@
 					break;
 				case NotificationPosition.TOP_RIGHT: // Top right should also be the default case
 				default:
-					if (latest != null) { // If a notification is already being displayed
-						int nx;
-						int ny;
-						latest.get_position(out nx, out ny);
-						x = nx;
-						y = ny + latest.get_child().get_allocated_height() + BUFFER_ZONE;
+					if (latest_exists) { // If a notification is already being displayed
+						x = existing_x;
+						y = existing_y + existing_height + BUFFER_ZONE;
 					} else { // This is the first nofication on the screen
 						x = (rect.x + rect.width) - NOTIFICATION_WIDTH;
 						x -= BUFFER_ZONE; // Don't touch edge of the screen
diff --git a/src/daemon/notifications/popup.vala b/src/daemon/notifications/popup.vala
index c11ce903..3d57e6e8 100644
--- a/src/daemon/notifications/popup.vala
+++ b/src/daemon/notifications/popup.vala
@@ -22,6 +22,8 @@ namespace Budgie.Notifications {
 
 		private uint expire_id { get; private set; }
 
+		public bool destroying { get; private set; default = false; }
+
 		public signal void Closed(NotificationCloseReason reason);
 
 		construct {
@@ -57,6 +59,7 @@ namespace Budgie.Notifications {
 		 * Destroy this notification popup.
 		 */
 		public void dismiss() {
+			this.destroying = true;
 			destroy();
 		}
 
-- 
2.34.1

